src/modules/silentpayments/main_impl.h

```diff
@@ -382,4 +382,379 @@ int secp256k1_silentpayments_recipient_create_labeled_spend_pubkey(const secp256
+                /* Serialize must succeed because the point was just loaded.
+                 *
+                 * Note: serialize will also fail if label_ge is the point at infinity, but we know
+                 * this cannot happen since we only hit this branch if tx_output != output_xonly.
+                 * Thus, we know that label_ge = tx_output_gej + output_negated_ge cannot be the
+                 * point at infinity.
+                 */
+                VERIFY_CHECK(ret && len == 33);
+                label_tweak = label_lookup(label33, label_context);
+                if (label_tweak != NULL) {
+                    found = 1;
+                    found_idx = i;
+                    break;
+                }
+            }
+        }
+        if (found) {
+            found_outputs[n_found]->output = *tx_outputs[found_idx];
+            secp256k1_scalar_get_b32(found_outputs[n_found]->tweak, &output_tweak_scalar);
+            if (label_tweak != NULL) {
+                found_outputs[n_found]->found_with_label = 1;
+                /* This is extremely unlikely to fail in that it can only really fail if label_tweak
+                 * is the negation of the shared secret tweak. But since both tweak and label_tweak are
+                 * created by hashing data, practically speaking this would only happen if an attacker
+                 * tricked us into using a particular label_tweak (deviating from the protocol).
+                 */
+                if (!secp256k1_ec_seckey_tweak_add(ctx, found_outputs[n_found]->tweak, label_tweak)) {
+                    secp256k1_scalar_clear(&scan_key_scalar);
+                    secp256k1_scalar_clear(&output_tweak_scalar);
+                    secp256k1_memclear_explicit(&shared_secret, sizeof(shared_secret));
+                    return 0;
+                }
+                secp256k1_pubkey_save(&found_outputs[n_found]->label, &label_ge);
+            } else {
+                found_outputs[n_found]->found_with_label = 0;
+                /* Set the label public key with an invalid public key value. */
+                memset(&found_outputs[n_found]->label, 0, sizeof(secp256k1_pubkey));
+            }
+            /* Reset everything for the next round of scanning. */
+            label_tweak = NULL;
+            n_found++;
+            k++;
+        } else {
+            break;
+        }
+    }
+    *n_found_outputs = n_found;
+
+    /* Leaking these values would break indistinguishability of the transaction, so clear them. */
+    secp256k1_scalar_clear(&scan_key_scalar);
+    secp256k1_scalar_clear(&output_tweak_scalar);
+    secp256k1_memclear_explicit(shared_secret, sizeof(shared_secret));
+    return 1;
+}
+
+int secp256k1_silentpayments_recipient_create_shared_secret(const secp256k1_context *ctx, unsigned char *shared_secret33, const unsigned char *scan_key32, const secp256k1_silentpayments_prevouts_summary *prevouts_summary) {
+    secp256k1_scalar rsk;
```

Comment:
naming nit: s/rsk/scan_key_scalar/ (after this, I'm pretty confident that all of the outdated or confusing scan/spend key variable names are fixed)